plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs

```diff
@@ -0,0 +1,118 @@
+-- | The CEK machine.
+-- Rules are the same as for the CK machine from "Language.PlutusCore.Evaluation.CkMachine",
+-- except we do not use substitution and use environments instead.
+-- The CEK machine relies on variables having
+-- 1. equal 'Unique's whenever they have equal string names
+-- 2. non-equal 'Unique's whenever they have non-equal string names
+-- I.e. 'Unique's are used instead of string names, so a renamer pass is required.
+-- This is for efficiency reasons.
+-- The CEK machine generates booleans along the way which might contain globally non-unique 'Unique's.
+-- This is not a problem as the CEK machines handles name capture by design.
+
+module PlutusCoreInterpreter.CekMachine
+    ( EvaluationResult (..)
+    , evaluateCek
+    , runCek
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Evaluation.MachineException (MachineError (..), MachineException (..))
+import           Language.PlutusCore.Evaluation.Result           (EvaluationResult (..))
+import           Language.PlutusCore.View
+import           PlutusPrelude
+
+import           Data.IntMap                                     (IntMap)
+import qualified Data.IntMap                                     as IntMap
+
+type Plain f = f TyName Name ()
+
+-- | Environments used by the CEK machine.
+-- Each row is a mapping from the 'Unique' representing a variable to the 'Value' the
+-- variable stands for and the environment the value was defined in.
+newtype Environment = Environment (IntMap (Plain Value, Environment))
```

Comment:
I'd be tempted to do `type Closure = (Plain Value, Environment)`.

Reply:
I think I'm going to make this an actual data type then.

Reply:
:+1: